common/lib/xmodule/xmodule/capa_base.py

```diff
@@ -1696,3 +1700,138 @@ def score_from_lcp(self):
+            return str(label)
+
+        # FIXME if there are 2 successive dropdowns, the 2nd is finding the question from the 1st one. Fix
+
+        xml_elems = [elem for elem, data in self.lcp.responder_answers.iteritems() if question_id in data]
+        assert len(xml_elems) == 1, (len(xml_elems), xml_elems, question_id, list(self.lcp.responder_answers.iteritems()))
+
+        # Get the element that probably contains the question text
+        questiontext_elem = xml_elems[0].getprevious()
+
+        # Go backwards, skip <description> and other responses, because they don't contain the question
+        skip_elems = registry.registered_tags() + ['description']
+        while questiontext_elem is not None and questiontext_elem.tag in skip_elems:
+            questiontext_elem = questiontext_elem.getprevious()
+
+        if questiontext_elem is not None and questiontext_elem.tag in ['p', 'label']:
+            return questiontext_elem.text
+        else:
+            return None
+
+    # FIXME delete course_key and block_key parameters (they are implicit)
+    # FIXME delete get_block parameter (unneeded?)
+    # FIXME implement filter by user_ids and match_string
+    # v1:
+    # def generate_report_data(self, course_key=None, block_key=None, get_block=None, user_ids=None, match_string=None):
+    # v2 (static):
+    # FIXME make static so that it can run in a smaller runtime, with ~~~~ only a CapaDescriptor, see e.g. CapaDescriptor.max_score about how to build such runtime. Maybe move to CapaDescriptor
+    @staticmethod
+    def generate_report_data(descriptor):
+        """
+        Return a list of student responses to this block in a readable way.
+        Each call returns a tuple like: ("username", {"Question": "2 + 2 equals how much?", "Answer": "Four"})
+        """
+
+        # import sys; sys.stdout = sys.__stdout__; import ipdb; ipdb.set_trace()
+
+        if True and "testing with static":
+            # import sys; sys.stdout = sys.__stdout__; import ipdb; ipdb.set_trace()
+            self = descriptor
+
+            # from xmodule.tests import DATA_DIR, get_test_system, get_test_descriptor_system
+            # self.xmodule_runtime = get_test_system()
+            # assert self.xmodule_runtime
+
+        if self.category != 'problem':
+            raise NotImplementedError()
+        # FIXME reimplement without using self.lcp (to make it work with the static one)
+        if False and 'customresponse' in [elem.tag for elem in self.lcp.responder_answers.keys()]:
+            raise NotImplementedError("Not implemented for custom response problems "
+                                      "(like drag&drop, chemical equations etc.)")
+
+        # FIXME move import (but if moved to be too early in the imports, then it fails)
+        from courseware.user_state_client import DjangoXBlockUserStateClient
+
+        # FIXME move to the right place (init just once)
+        user_state_client = DjangoXBlockUserStateClient()
```

Comment:
@clemente @xitij2000  XBlocks are not supposed to know about django or the specifics of the LMS, so this should not be here. I had proposed that the celery task (in #18013) would call initialize `DjangoXBlockUserStateClient()` and pass it in to this function as a parameter, by calling `Block.generate_report_data(user_state_client)`. Is there any reason for not doing it that way?

Reply:
@bradenmacdonald Yeah that makes sense. I assumed though that the user_state_client method was just here for testing. I think I'll refactor my code to pass this along so @clemente  can test the report generation by actually generating reports. 

@clemente When I push that change could you rebase your changes on my PR?

Reply:
@bradenmacdonald ok, that's a easier function signature. For reference, the initial function signature that I found in the document was `def generate_report_data(self, course_key=None, block_key=None, get_block=None, user_ids=None, match_string=None)`, which seems to have too many parameters.

@xitij2000 I rebased and pushed. Could you add some instructions to your PR or here about how to run your task in a devstack?

Reply:
@clemente You can just use the LMS to generate the report for a particular block. 
Or if you wan to do so from the shell you could just call ``ProblemResponses._build_student_data(user_id:int, course_id:CourseKey, problem_location:str)`` and it will return an array of responses. 

Note that you should also accept a ``limit_responses`` parameter in ``generate_report_data`` which will limit how many responses will be returned.